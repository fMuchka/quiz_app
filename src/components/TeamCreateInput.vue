<template>
  <div 
    class="input-wrapper"
    >
      <label
        for="team-input">

       Team {{ teamColor }}: 
      
       </label>

      <input 
        name="team-input"
        type="text"
        />

      <div 
        class="plus"
        v-on:click="emitNewTeam"
      />
  </div>
</template>

<script>
export default {
    computed: {
        teamColor: {
            get() {
                return 'PLACEHOLDER';
            }
        }
    },
    methods: {
        emitNewTeam(){
           this.$emit('new-team');
        }
    }
}
</script>

<style>
.input-wrapper{
    display: flex;
} 
.plus{
    width:50px;
    height:50px;
    display:inline-block;

    background:
    linear-gradient(#000,#000) top left,
    linear-gradient(#000,#000) top right,
    linear-gradient(#000,#000) bottom left,
    linear-gradient(#000,#000) bottom right;
    background-size: calc(50% - 1px) calc(50% - 1px); /*thickness = 2px (2*1px) */
    background-repeat:no-repeat;
    border:10px solid #000; /*length = 30px (50px - 2x10px) */
    box-sizing:border-box;
    border-radius: 50%;
}
.plus:hover{
    cursor: pointer;
}
</style>