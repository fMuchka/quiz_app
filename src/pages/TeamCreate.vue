<template>
  <div>
    <button id="addButton" v-on:click="addNewTeam"></button>

    <team-create-input
      v-for="(team, index) in teamList"
      :team="teamList[index]"
      :key="team.id"
      :index="index"
      @remove="teamList.splice(index, 1)"
    />
  </div>
</template>

<script>
import TeamCreateInput from "../components/TeamCreateInput.vue";

export default {
  components: {
    TeamCreateInput,
  },

  data() {
    return {
      nextTeamId: 1,
      teamList: [],
    };
  },

  computed: {},

  methods: {
    addNewTeam() {
      const freeIndex = this.$store.getters.nextColorIndex;
      const colorListLength = this.$store.getters.colorListLength;
      if (freeIndex < colorListLength) {
        this.teamList.push({
          id: this.nextTeamId++,
          label: "",
          color: this.$store.getters.color,
        });

        this.$store.commit("incrementColorIndex");
        this.$store.commit("changeColorStatus", {
          index: freeIndex,
          status: true,
        });
      }
    },
  },
};
</script>
<style scoped>
#addButton {
  width: 50px;
  height: 50px;
  display: inline-block;

  background: linear-gradient(#000, #000) top left,
    linear-gradient(#000, #000) top right,
    linear-gradient(#000, #000) bottom left,
    linear-gradient(#000, #000) bottom right;
  background-size: calc(50% - 1px) calc(50% - 1px); /*thickness = 2px (2*1px) */
  background-repeat: no-repeat;
  border: 10px solid #000; /*length = 30px (50px - 2x10px) */
  box-sizing: border-box;
  border-radius: 50%;
}
#addButton:hover {
  cursor: pointer;
}
</style>
