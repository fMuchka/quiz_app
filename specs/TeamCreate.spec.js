import Vuex from "vuex";
import { mount, createLocalVue } from "@vue/test-utils";

import TeamCreate from "../src/pages/TeamCreate.vue";
import TeamCreateInput from "../src/components/TeamCreateInput.vue"

const localVue = createLocalVue();
localVue.use(Vuex);

describe("TeamCreate Page:", () => {
  let store;

  // had to simulate store, otherwise would return error of undefined
  // guess it makes sense...
  beforeEach(() => {
      store = new Vuex.Store({
        state: {
          teams: [],
          colorList: [
            { id: 'red', label: 'Červený', used: false },
            { id: 'blue', label: 'Modrý', used: false },
            { id: 'green', label: 'Zelený', used: false },
            { id: 'yellow', label: 'Žlutý', used: false },
            { id: 'purple', label: 'Fialový', used: false },
            { id: 'brown', label: 'Hnědý', used: false },
          ]
        },
        mutations: {
          pushTeam(state, team) {
            state.teams.push(team);
          },
          changeColorStatus(state, input) {
            const { id, status } = input;
            const colorList = state.colorList;

            for (let i = 0; i < colorList.length; i++) {
              if (colorList[i].id === id) {
                colorList[i].used = status;
                break;
              }
            }
          },
        },
        getters: {
          color: (state) => {
            const colors = state.colorList;
            const freeIndex = store.getters.nextColorIndex;

            if (freeIndex === false) {
              // no colors left
              return false;
            }

            if (freeIndex < colors.length) {
              return colors[freeIndex];
            }

          },
          nextColorIndex: (state) => {
            const colorList = state.colorList;

            for (let i = 0; i < colorList.length; i++) {
              if (colorList[i].used === false) {
                return i;
              }
            }
            return false;
          },
          colorListLength(state) {
            return state.colorList.length;
          }
        }
      });
    });

  it("html structure is correct", () => {
    const wrapper = mount(TeamCreate, {
      localVue,
      store,
    });
  
    expect(wrapper.element).toMatchSnapshot();
  });

  it("handles removing team inputs", () => {
    const wrapper = mount(TeamCreate, {
      localVue,
      store,
    }); 

    expect(true).toBe(false);
  });

  describe("Add button:", () => {
    it("creates team input on click", () => {
      const wrapper = mount(TeamCreate, {
        localVue,
        store
      });

      const button = wrapper.find("button");
      const addSpy = jest.spyOn(wrapper.vm, "addTeamInput");

      expect(wrapper.vm.teamList.length).toBe(0);
      button.trigger('click');
      expect(addSpy).toHaveBeenCalled();
      expect(wrapper.vm.teamList.length).toBe(1);
  });

    it("disables when all inputs created", () => {
      const wrapper = mount(TeamCreate, {
        localVue,
        store
      });

      const button = wrapper.find("button");
      expect(wrapper.vm.isAddDisabled).toBe(false);

      const colorListLength = store.getters.colorListLength;

      for (let i = 0; i < colorListLength; i++) {
          button.trigger('click');
      }
      
      expect(wrapper.vm.isAddDisabled).toBe(true);
    });
  });
  describe("TeamCreateInput component", () => {

  });
});
