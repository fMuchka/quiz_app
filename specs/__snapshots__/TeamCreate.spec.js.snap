// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamCreate Page: html structure is correct 1`] = `
<div
  id="pageWrapper"
>
  <label
    for="addButton"
  >
    Přidat tým
  </label>
   
  <button
    id="addButton"
  >
    
    +
  
  </button>
   
</div>
`;
